src/OpenSage.Game/Logic/Object/Draw/W3dModelDraw.cs

```diff
@@ -119,6 +119,20 @@ public override void UpdateConditionState(BitArray<ModelConditionFlag> flags)
             }
 
             SetActiveConditionState(bestConditionState);
+
+            bestConditionState.WeaponMuzzleFlashes.ForEach(x => {
+                var visible = flags.Get(ModelConditionFlag.FiringA);
```

Comment:
This is wrong. FiringA is used when firing from the primary weapon slot. FiringB from the secondary etc. So you'd have to check that.
I would also recommend to move this out of this function. Maybe make it event driven and let the muzzle flash object subscribe to it when needed.
Also usually the muzzle flash is usually only appearing in the default, as the others are basically overrides. So you should first check the current state and if not found check the default.

Another thing I noticed (which is not related to your code) is that EA's SAGE games choose the current condition state in quite a different manner. They pick the first state which has all the bits set, regardless if there is another more relevant state coming after that. It might not be as intuitive but should offer better performance.